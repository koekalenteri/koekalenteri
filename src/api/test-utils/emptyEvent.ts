import type { ConfirmedEvent } from '../../types'

export const emptyEvent: ConfirmedEvent = {
  classes: [{ class: 'ALO' }],
  contactInfo: {
    official: {
      name: 'Teemu Toimitsija',
      email: 'official@example.com',
      phone: '040-official',
    },
    secretary: {
      name: 'Siiri Sihteeri',
      email: 'secretary@example.com',
      phone: '040-secretary',
    },
  },
  cost: 123,
  costMember: 123,
  createdAt: new Date('1989-01-01'),
  createdBy: 'test',
  description: 'test',
  endDate: new Date('1990-01-01'),
  entries: 0,
  entryEndDate: new Date('1990-01-01'),
  entryStartDate: new Date('1990-01-01'),
  eventType: 'test',
  id: 'test',
  judges: [123],
  kcId: 123456,
  location: 'test',
  modifiedAt: new Date('1989-01-02'),
  modifiedBy: 'test',
  name: 'test',
  official: {
    id: '1',
    name: 'Teemu Toimitsija',
  },
  priority: ['member'],
  secretary: {
    id: '2',
    name: 'Siiri Sihteeri',
  },
  organizer: {
    id: '1',
    name: 'Suomen Noutajakoirajärjestö ry',
  },
  places: 10,
  startDate: new Date('1990-01-01'),
  state: 'confirmed' as const,
}
