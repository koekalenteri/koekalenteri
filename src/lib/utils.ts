import type {
  DeepPartial,
  EventState,
  JsonValue,
  PublicContactInfo,
  PublicDogEvent,
  RegistrationDate,
  RegistrationTime,
} from '../types'

import { eachDayOfInterval, subDays } from 'date-fns'
import { diff } from 'deep-object-diff'

import { zonedEndOfDay, zonedStartOfDay } from '../i18n/dates'

type EventVitals = Partial<Pick<PublicDogEvent, 'startDate' | 'endDate' | 'entryStartDate' | 'entryEndDate' | 'state'>>

export const isValidForEntry = (state?: EventState) => !['draft', 'tentative', 'cancelled'].includes(state ?? '')

export const isEventUpcoming = ({ startDate }: EventVitals, now = new Date()) => !!startDate && startDate > now

export const isEntryUpcoming = ({ entryStartDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate && entryStartDate > now && (isValidForEntry(state) || state === 'tentative')

export const isEntryOpen = ({ entryStartDate, entryEndDate, state }: EventVitals, now = new Date()) =>
  !!entryStartDate &&
  !!entryEndDate &&
  zonedStartOfDay(entryStartDate) <= zonedEndOfDay(now) &&
  zonedEndOfDay(entryEndDate) >= zonedEndOfDay(now) &&
  isValidForEntry(state)

export const isEntryClosing = (event: EventVitals, now = new Date()) =>
  !!event.entryEndDate &&
  isEntryOpen(event, now) &&
  subDays(event.entryEndDate, 7) <= zonedEndOfDay(now) &&
  isValidForEntry(event.state)

export const isEntryClosed = ({ startDate, entryEndDate }: EventVitals, now = new Date()) =>
  !!startDate && !!entryEndDate && zonedEndOfDay(entryEndDate) < now && zonedStartOfDay(startDate) > now

export const isEventOngoing = ({ startDate, endDate, state }: EventVitals, now = new Date()) =>
  !!startDate &&
  !!endDate &&
  zonedStartOfDay(startDate) <= zonedEndOfDay(now) &&
  zonedEndOfDay(endDate) >= zonedEndOfDay(now) &&
  isValidForEntry(state) &&
  state !== 'confirmed'

export const isEventOver = ({ endDate }: EventVitals, now = new Date()) => !!endDate && zonedEndOfDay(endDate) < now

export const eventDates = (event?: Pick<PublicDogEvent, 'classes' | 'startDate' | 'endDate'> | null) => {
  if (!event) return []
  return event.classes.length
    ? uniqueDate(event.classes.map((c) => c.date ?? event.startDate))
    : eachDayOfInterval({ start: event.startDate, end: event.endDate })
}

export const uniqueClasses = (event?: Pick<PublicDogEvent, 'classes'> | null) =>
  unique((event?.classes ?? []).map((c) => c.class))

export const placesForClass = (
  event: DeepPartial<Pick<PublicDogEvent, 'places' | 'classes'>> | undefined | null,
  cls: string
) => {
  if (!event) {
    return 0
  }

  return (
    (event.classes ?? []).filter((c) => c.class === cls).reduce((acc, cur) => acc + (Number(cur.places) || 0), 0) ||
    Number(event.places) ||
    0
  )
}

export const uniqueClassDates = (event: PublicDogEvent, cls: string) =>
  cls === event.eventType
    ? eventDates(event)
    : uniqueDate(event.classes.filter((c) => c.class === cls).map((c) => c.date ?? event.startDate))

export const registrationDates = (event: PublicDogEvent, times: RegistrationTime[], cls?: string | null) =>
  (cls ? uniqueClassDates(event, cls) : eventDates(event)).flatMap<RegistrationDate>((date) =>
    times.map((time) => ({ date, time }))
  )

export const unique = <T = string>(arr: T[]): T[] => arr.filter((c, i, a) => a.indexOf(c) === i)

export const uniqueFn = <T>(arr: T[], cmp: (a: T, b: T) => boolean): T[] =>
  arr.filter((c, i, a) => a.findIndex((f) => cmp(f, c)) === i)

export const uniqueDate = (arr: Date[]) => [...new Set<number>(arr.map((d) => d.valueOf()))].map((v) => new Date(v))

const DATE_RE = /^\d{4}-(?:0[1-9]|1[0-2])-(?:[0-2][1-9]|[1-3]0|3[01])$/
const TIME_RE = /^(?:[0-1]\d|2[0-3])(?::[0-6]\d)(?::[0-6]\d)?(?:\.\d{3})?(?:[+-][0-2]\d:[0-5]\d|Z)?$/

export const isDateString = (value: unknown): value is string => {
  if (typeof value !== 'string') {
    return false
  }
  const [date, time] = value.split('T')
  return DATE_RE.test(date) && TIME_RE.test(time)
}

function dateReviver(_key: string, value: JsonValue): JsonValue | Date {
  if (isDateString(value)) {
    const dateObj = new Date(value)
    if (!isNaN(+dateObj)) {
      return dateObj
    }
  }
  return value
}

export const parseJSON = (json: string, reviveDates: boolean = true) => {
  if (!json) return undefined

  const reviver = reviveDates ? dateReviver : undefined

  return JSON.parse(json, reviver)
}

export type AnyObject = Record<string, any>
export type EmptyObject = Record<string, never>
export type Entries<T> = {
  [K in keyof T]: [K, T[K]]
}[keyof T][]

export const isEmpty = (o: AnyObject): o is EmptyObject => Object.keys(o).length === 0

export const isObject = (o: unknown): o is AnyObject =>
  o !== null && Object.prototype.toString.call(o) === '[object Object]'

export const isEmptyObject = (o: unknown): o is EmptyObject => isObject(o) && isEmpty(o)

export const hasChanges = (a: object | undefined | null, b: object | undefined | null): boolean =>
  !isEmptyObject(diff(a ?? {}, b ?? {}))

export const clone = <T extends AnyObject>(a: T): T => ({ ...a })

export const merge = <T extends AnyObject>(a: T, b: DeepPartial<T>): T => {
  const result = isObject(a) ? clone(a) : ({} as T)
  if (!isObject(b)) {
    return result
  }
  for (const [key, value] of Object.entries(b) as Entries<T>) {
    if (isObject(value)) {
      const old = result[key]
      result[key] = isObject(old) ? merge(old, value) : value
    } else {
      result[key] = value
    }
  }
  return result
}

// This function merges a patch object into a base object, but only creates new objects for paths that actually change
export const patchMerge = <T extends AnyObject, P extends DeepPartial<T>>(base: T, patch: P): T & P => {
  if (!isObject(base) || !isObject(patch)) return patch as T & P

  let changed = false
  const result = { ...base } as T & P
  for (const key of Object.keys(patch)) {
    const merged = patchMerge(base[key], (patch as any)[key])
    if (merged !== base[key]) {
      ;(result as any)[key] = merged
      changed = true
    }
  }

  return changed ? result : (base as T & P)
}

export const printContactInfo = (info?: PublicContactInfo) =>
  [info?.name, info?.phone, info?.email].filter(Boolean).join(', ')
