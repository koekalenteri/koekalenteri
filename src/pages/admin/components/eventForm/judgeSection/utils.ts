import type { DeepPartial, EventClass, EventType, Judge, PublicJudge } from '../../../../../types'
import type { PartialEvent } from '../../EventForm'

import { isSameDay } from 'date-fns'

export type PartialPublicJudge = Partial<PublicJudge>
export type PartialPublicJudgeValue = PartialPublicJudge | PartialPublicJudge[]

export const filterJudges = (
  judges: Judge[],
  eventJudges: PublicJudge[],
  id: number | undefined,
  eventType?: EventType
) =>
  judges
    .filter((j) => !eventType?.official || j.eventTypes.includes(eventType.eventType))
    .filter((j) => j.id === id || !eventJudges.find((ej) => ej.id === j.id))

export const hasJudge = (c: DeepPartial<EventClass>, id?: number) =>
  Array.isArray(c.judge) ? c.judge.find((j) => j.id === id) : c.judge?.id === id

export const filterClassesByJudgeId = (classes?: EventClass[], id?: number) => classes?.filter((c) => hasJudge(c, id))

export const toArray = (j?: PartialPublicJudge): PartialPublicJudge[] => (j ? [j] : [])
export const makeArray = (j?: PartialPublicJudgeValue) => (Array.isArray(j) ? [...j] : toArray(j))
export const selectJudge = (j?: PartialPublicJudgeValue, judge?: PublicJudge): PartialPublicJudge[] => {
  const a = makeArray(j)
  if (judge) {
    const index = a.findIndex((cj) => cj.id === judge.id)
    if (index === -1) {
      a.push(judge)
    } else {
      a[index] = judge
    }
  }
  return a
}

export const removeJudge = (j?: PartialPublicJudgeValue, id?: number): PartialPublicJudge[] => {
  const a = makeArray(j)
  return a.filter((cj) => cj.id !== id)
}

export const updateJudge = (
  event: PartialEvent,
  id: number | undefined,
  judge: PublicJudge | undefined,
  values?: DeepPartial<EventClass>[]
) => {
  const isSelected = (c: DeepPartial<EventClass>) =>
    values?.find((v) => event && isSameDay(v.date ?? event.startDate, c.date ?? event.startDate) && v.class === c.class)
  return event.classes?.map((c) => ({
    ...c,
    judge: isSelected(c) ? selectJudge(c.judge, judge) : removeJudge(c.judge, id),
  }))
}
