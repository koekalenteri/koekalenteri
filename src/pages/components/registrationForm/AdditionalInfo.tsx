import type { ChangeEventHandler } from 'react'
import type { Registration } from '../../../types'

import { useCallback, useEffect, useState } from 'react'
import { useTranslation } from 'react-i18next'
import TextField from '@mui/material/TextField'

import useDebouncedCallback from '../../../hooks/useDebouncedCallback'
import CollapsibleSection from '../CollapsibleSection'

interface Props {
  readonly disabled?: boolean
  readonly notes?: string
  readonly onChange?: (props: Partial<Registration>) => void
  readonly onOpenChange?: (value: boolean) => void
  readonly open?: boolean
}

export function AdditionalInfo({ disabled, notes, onChange, onOpenChange, open }: Props) {
  const { t } = useTranslation()
  const [value, setValue] = useState(notes ?? '')

  const dispatchChange = useDebouncedCallback((notes: string) => onChange?.({ notes }))

  const handleChange = useCallback<ChangeEventHandler<HTMLTextAreaElement>>(
    (e) => {
      const newValue = e.target.value
      setValue(newValue)
      dispatchChange(newValue)
    },
    [dispatchChange]
  )

  useEffect(() => {
    setValue(notes ?? '')
  }, [notes])

  return (
    <CollapsibleSection title={t('registration.notes')} open={open} onOpenChange={onOpenChange}>
      <TextField
        disabled={disabled}
        label={t('registration.notes')}
        multiline
        name="notes"
        onChange={handleChange}
        rows={4}
        sx={{ width: '100%' }}
        value={value}
      />
    </CollapsibleSection>
  )
}
