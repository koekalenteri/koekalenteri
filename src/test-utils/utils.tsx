import type { RenderOptions, RenderResult } from '@testing-library/react'
import type { Options } from '@testing-library/user-event/dist/types/options'
import type { UserEvent } from '@testing-library/user-event/dist/types/setup/setup'
import type { RouteObject, RouterInit } from 'react-router'
import type { RecoilValue } from 'recoil'

import { act, useEffect } from 'react'
import { createMemoryRouter, createRoutesFromElements, RouterProvider } from 'react-router'
import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import mediaQuery from 'css-mediaquery'
import { useRecoilValue } from 'recoil'

/**
 * Abstraction to avoid re-writing all tests for the time being
 * @see https://github.com/remix-run/react-router/blob/main/packages/react-router/__tests__/data-memory-router-test.tsx
 */
export function DataMemoryRouter({
  basename,
  children,
  hydrationData,
  initialEntries,
  initialIndex,
  routes,
}: {
  readonly basename?: RouterInit['basename']
  readonly children?: React.ReactNode | React.ReactNode[]
  readonly fallbackElement?: React.ReactNode
  readonly hydrationData?: RouterInit['hydrationData']
  readonly initialEntries?: string[]
  readonly initialIndex?: number
  readonly routes?: RouteObject[]
}) {
  const router = createMemoryRouter(routes ?? createRoutesFromElements(children), {
    basename,
    hydrationData,
    initialEntries,
    initialIndex,
  })
  return <RouterProvider router={router} />
}

export const flushPromises = async (timers: boolean = true) => {
  for (let i = 0; i <= 7; i++) {
    await act(async () => {
      if (timers) jest.runOnlyPendingTimers()
      await Promise.resolve()
    })
  }
}

export const createMatchMedia =
  (width: number) =>
  (query: string): MediaQueryList => ({
    matches: mediaQuery.match(query, { width }),
    media: query,
    onchange: null,
    addListener: jest.fn(), // deprecated
    removeListener: jest.fn(), // deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })

export function RecoilObserver<T>({
  node,
  onChange,
}: {
  readonly node: RecoilValue<T>
  readonly onChange: (value: T) => void
}) {
  const value = useRecoilValue(node)
  useEffect(() => onChange(value), [onChange, value])
  return null
}

export function renderWithUserEvents(
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
  userEventOptions?: Options
): RenderResult & { user: UserEvent } {
  return {
    user: userEvent.setup(userEventOptions),
    ...render(ui, options),
  }
}
