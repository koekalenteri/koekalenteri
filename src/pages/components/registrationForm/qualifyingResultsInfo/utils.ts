import type { EventResultRequirement, EventResultRequirements, EventResultRequirementsByDate } from '../../../../rules'
import type { ManualTestResult, QualifyingResult } from '../../../../types'

import { nanoid } from 'nanoid'

import { unique } from '../../../../lib/utils'
import { objectContains } from '../validation'

const asArray = (v: EventResultRequirements | EventResultRequirement) => (Array.isArray(v) ? v : [v])

function findFirstMissing(requirements: EventResultRequirementsByDate | undefined, results: QualifyingResult[]) {
  if (!requirements || typeof requirements.rules === 'function') {
    return []
  }
  for (const rule of requirements.rules) {
    for (const opt of asArray(rule)) {
      const { count, ...rest } = opt
      if (results.filter((r) => objectContains(r, rest)).length < count) {
        return rest
      }
    }
  }
}

export function availableTypes(requirements?: EventResultRequirementsByDate, eventType?: string) {
  if (!requirements || typeof requirements.rules === 'function') {
    switch (eventType) {
      case 'NOME-B SM':
        return ['NOME-B']
      case 'NOME-A SM':
        return ['NOME-A', 'NOME-A KV']
      case 'NOWT SM':
        return ['NOWT']
    }
    return []
  }
  return unique(requirements.rules.flatMap((rule) => asArray(rule).map((opt) => opt.type)))
}

const availableSMResults = (type?: string, eventType?: string, hasKVA?: boolean) => {
  switch (eventType) {
    case 'NOME-B SM':
      return [hasKVA ? undefined : 'FI KVA-B', 'VOI1', 'VOI2', 'VOI3'].filter(Boolean)
    case 'NOWT SM':
      return [hasKVA ? undefined : 'FI KVA-WT', 'VOI1', 'VOI2', 'VOI3'].filter(Boolean)
    case 'NOME-A SM':
      if (type === 'NOME-A')
        return [hasKVA ? undefined : 'FI KVA-FT', 'A1 CERT', 'A1 RES-CERT', 'A1', 'A2', 'A3'].filter(Boolean)

      if (type === 'NOME-A KV') return ['EXC CACIT', 'EXC RES-CACIT', 'EXC', 'VG', 'G']
  }
  return []
}

export function availableResults(
  requirements?: EventResultRequirementsByDate,
  type?: string,
  eventType?: string,
  existingResults?: ManualTestResult[]
) {
  if (!requirements || typeof requirements.rules === 'function') {
    if (eventType?.includes(' SM')) {
      const hasKVA = existingResults?.some((r) => r.result.startsWith('FI KVA'))

      return availableSMResults(type, eventType, hasKVA)
    }
    return []
  }
  return unique(
    requirements.rules.flatMap((rule) =>
      asArray(rule)
        .filter((opt) => !type || opt.type === type)
        .map((opt) => (opt.cert ? 'CERT' : opt.result))
    )
  )
}

export function createMissingResult(
  requirements: EventResultRequirementsByDate | undefined,
  results: Array<QualifyingResult | ManualTestResult>,
  regNo: string
): ManualTestResult {
  const rule = findFirstMissing(requirements, results)
  return {
    id: nanoid(10),
    regNo,
    date: new Date(),
    official: false,
    qualifying: true,
    type: '',
    judge: '',
    location: '',
    result: '',
    class: '',
    ...rule,
  }
}

export function resultBorderColor(qualifying: boolean | undefined) {
  if (qualifying === true) {
    return 'success.light'
  }
  if (qualifying === false) {
    return 'error.main'
  }
}

export function getResultId(result: ManualTestResult | QualifyingResult) {
  if ('id' in result) {
    return result.id
  }
  return result.type + result.class + result.date.toISOString()
}
