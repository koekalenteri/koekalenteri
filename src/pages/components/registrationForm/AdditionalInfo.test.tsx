import { render, screen } from '@testing-library/react'

import { flushPromises, renderWithUserEvents } from '../../../test-utils/utils'

import { AdditionalInfo } from './AdditionalInfo'

describe('AdditionalInfo', () => {
  beforeAll(() => jest.useFakeTimers())
  afterEach(() => jest.runOnlyPendingTimers())
  afterAll(() => jest.useRealTimers())

  it('should render with minimal info', () => {
    const { container } = render(<AdditionalInfo />)
    expect(container).toMatchSnapshot()
  })

  it('should render with text', () => {
    const { container } = render(<AdditionalInfo notes="test" />)
    expect(container).toMatchSnapshot()
  })

  it('should rerender with new value', () => {
    const { container, rerender } = render(<AdditionalInfo notes="test a" />)
    rerender(<AdditionalInfo notes="changed notes" />)
    expect(container).toMatchSnapshot()
  })

  it('should call onChange', async () => {
    const onChange = jest.fn()
    const { user } = renderWithUserEvents(<AdditionalInfo notes="test" onChange={onChange} />, undefined, {
      advanceTimers: jest.advanceTimersByTime,
    })

    await flushPromises()
    expect(onChange).not.toHaveBeenCalled()

    const input = screen.getByRole('textbox')

    await user.clear(input)
    await flushPromises()
    expect(onChange).toHaveBeenLastCalledWith({ notes: '' })

    onChange.mockReset()

    await user.type(input, 'testing notes')
    await flushPromises()
    expect(onChange).toHaveBeenLastCalledWith({ notes: 'testing notes' })
  })
})
