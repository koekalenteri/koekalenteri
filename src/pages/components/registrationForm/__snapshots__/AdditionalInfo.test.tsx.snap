// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalInfo should render with minimal info 1`] = `
<div>
  <div
    class="MuiBox-root css-l530ik"
  >
    <button
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-19u30f2-MuiButtonBase-root-MuiIconButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
        />
      </svg>
    </button>
    <div
      class="MuiBox-root css-1qbrvqz"
    >
      <div
        class="MuiBox-root css-1jw9kr7"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
        >
          registration.notes
        </p>
        <p
          class="MuiFormHelperText-root css-gorocw-MuiFormHelperText-root"
        />
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-wz2cs5-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiBox-root css-e1h6et"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1q964xs-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for=":r0:"
                  id=":r0:-label"
                >
                  registration.notes
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-9yqzd6-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1eiam5b-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r0:"
                    name="notes"
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  />
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1eiam5b-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-1pbc52w"
                    >
                      <span>
                        registration.notes
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdditionalInfo should render with text 1`] = `
<div>
  <div
    class="MuiBox-root css-l530ik"
  >
    <button
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-19u30f2-MuiButtonBase-root-MuiIconButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
        />
      </svg>
    </button>
    <div
      class="MuiBox-root css-1qbrvqz"
    >
      <div
        class="MuiBox-root css-1jw9kr7"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
        >
          registration.notes
        </p>
        <p
          class="MuiFormHelperText-root css-gorocw-MuiFormHelperText-root"
        />
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-wz2cs5-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiBox-root css-e1h6et"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-18pjc51-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for=":r1:"
                  id=":r1:-label"
                >
                  registration.notes
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-9yqzd6-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1eiam5b-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r1:"
                    name="notes"
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  >
                    test
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1eiam5b-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-kg5swy"
                    >
                      <span>
                        registration.notes
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AdditionalInfo should rerender with new value 1`] = `
<div>
  <div
    class="MuiBox-root css-l530ik"
  >
    <button
      class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorPrimary MuiIconButton-sizeSmall css-19u30f2-MuiButtonBase-root-MuiIconButton-root"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-20bmp1-MuiSvgIcon-root"
        data-testid="KeyboardArrowDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6z"
        />
      </svg>
    </button>
    <div
      class="MuiBox-root css-1qbrvqz"
    >
      <div
        class="MuiBox-root css-1jw9kr7"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-1edfpdg-MuiTypography-root"
        >
          registration.notes
        </p>
        <p
          class="MuiFormHelperText-root css-gorocw-MuiFormHelperText-root"
        />
      </div>
      <div
        class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-entered css-wz2cs5-MuiCollapse-root"
        style="min-height: 0px;"
      >
        <div
          class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
        >
          <div
            class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
          >
            <div
              class="MuiBox-root css-e1h6et"
            >
              <div
                class="MuiFormControl-root MuiTextField-root css-136h9lv-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-18pjc51-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="true"
                  for=":r2:"
                  id=":r2:-label"
                >
                  registration.notes
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-9yqzd6-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <textarea
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1eiam5b-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r2:"
                    name="notes"
                    rows="4"
                    style="height: 0px; overflow: hidden;"
                  >
                    changed notes
                  </textarea>
                  <textarea
                    aria-hidden="true"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1eiam5b-MuiInputBase-input-MuiOutlinedInput-input"
                    readonly=""
                    style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                    tabindex="-1"
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-kg5swy"
                    >
                      <span>
                        registration.notes
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
